AArch64 CI tests: qemu hits memory limit and fails with SIMD tests enabled

The AArch64 CI test that runs using QEMU fails consistently for PR #1871 and the reasons are not clear - here's the relevant excerpt from the log:
```
2020-06-13T16:29:49.3730503Z test wast::Cranelift::spec::simd::simd_i32x4_cmp ... ok
2020-06-13T16:29:57.9345959Z test wast::Cranelift::spec::simd::simd_i8x16_sat_arith ... ignored
2020-06-13T16:30:08.5287111Z test wast::Cranelift::spec::simd::simd_lane ... ignored
2020-06-13T16:30:15.8261749Z test wast::Cranelift::spec::simd::simd_load ... ignored
2020-06-13T16:49:23.7624987Z error: test failed, to rerun pass '-p wasmtime-cli --test all'
2020-06-13T16:49:23.7648421Z 
2020-06-13T16:49:23.7651248Z Caused by:
2020-06-13T16:49:23.7664954Z   process didn't exit successfully: `/home/runner/qemu/bin/qemu-aarch64 -L /usr/aarch64-linux-gnu /home/runner/work/wasmtime/wasmtime/target/aarch64-unknown-linux-gnu/release/deps/all-0af4aa3748ec4770` (signal: 9, SIGKILL: kill)
2020-06-13T16:49:24.0613948Z ##[error]Process completed with exit code 101.
2020-06-13T16:49:25.4620071Z Post job cleanup.
```
I have reproduced the test environment locally using the following commands:
```
rm -rf qemu-5.0.0 ${HOME}/qemu
curl https://download.qemu.org/qemu-5.0.0.tar.xz | tar xJf -
cd qemu-5.0.0
./configure --target-list=aarch64-linux-user --prefix=${HOME}/qemu --disable-tools --disable-slirp --disable-fdt --disable-capstone --disable-docs
make -j$(nproc) install
cd ..
RUSTFLAGS="-D warnings" \
  CARGO_INCREMENTAL=0 \
  CARGO_PROFILE_DEV_DEBUG=1 \
  CARGO_PROFILE_TEST_DEBUG=1 \
  CARGO_BUILD_TARGET=aarch64-unknown-linux-gnu \
  CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_RUNNER="${HOME}/qemu/bin/qemu-aarch64 -L /usr/aarch64-linux-gnu" \
  CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER=aarch64-linux-gnu-gcc \
  RUST_BACKTRACE=1 \
  cargo test \
  --features test-programs/test_programs \
  --release \
  --all \
  --exclude lightbeam \
  --exclude peepmatic \
  --exclude peepmatic-automata \
  --exclude peepmatic-fuzzing \
  --exclude peepmatic-macro \
  --exclude peepmatic-runtime \
  --exclude peepmatic-test \
  --exclude wasmtime-fuzz
```
However, I don't experience any test failures. In addition to that, I don't see any issues either when I run the test natively in an AArch64 environment. In that case the list of commands can be simplified to:
```
cargo test --release --all --exclude lightbeam
```
Note that the `--features test-programs/test_programs` parameter is omitted because it requires `rust-lld`, which appears not to be a part of the native AArch64 toolchain.

This issue has also been discussed in PR #1802.

cc @cfallin

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -327,6 +327,10 @@ jobs:
         upcase=$(echo ${{ matrix.target }} | awk '{ print toupper($0) }' | sed 's/-/_/g')
         echo ::set-env name=CARGO_TARGET_${upcase}_RUNNER::$HOME/qemu/bin/${{ matrix.qemu }}
         echo ::set-env name=CARGO_TARGET_${upcase}_LINKER::${{ matrix.gcc }}
+
+        # See comments in the source for why we enable this during QEMU
+        # emulation.
+        echo ::set-env name=WASMTIME_TEST_NO_HOG_MEMORY::1
       if: matrix.target != '' && matrix.os == 'ubuntu-latest'
 
     # Install wasm32-wasi target in order to build wasi-common's integration

--- a/build.rs
+++ b/build.rs
@@ -181,6 +181,7 @@ fn ignore(testsuite: &str, testname: &str, strategy: &str) -> bool {
         },
         "Cranelift" => match (testsuite, testname) {
             ("simd", "simd_address") => return false,
+            ("simd", "simd_align") => return false,
             ("simd", "simd_bitwise") => return false,
             ("simd", "simd_boolean") => return false,
             ("simd", "simd_i8x16_cmp") => return false,

--- a/tests/all/wast.rs
+++ b/tests/all/wast.rs
@@ -25,6 +25,19 @@ fn run_wast(wast: &str, strategy: Strategy) -> anyhow::Result<()> {
         .strategy(strategy)?
         .cranelift_debug_verifier(true);
 
+    // By default we'll allocate huge chunks (6gb) of the address space for each
+    // linear memory. This is typically fine but when we emulate tests with QEMU
+    // it turns out that it causes memory usage to balloon massively. Leave a
+    // knob here so on CI we can cut down the memory usage of QEMU and avoid the
+    // OOM killer.
+    //
+    // Locally testing this out this drops QEMU's memory usage running this
+    // tests suite from 10GiB to 600MiB. Prevoiusly we saw that crossing the
+    // 10GiB threshold caused our processes to get OOM killed on CI.
+    if std::env::var("WASMTIME_TEST_NO_HOG_MEMORY").is_ok() {
+        cfg.static_memory_maximum_size(0);
+    }
+
     let store = Store::new(&Engine::new(&cfg));
     let mut wast_context = WastContext::new(store);
     wast_context.register_spectest()?;

--- a/tests/all/wast.rs
+++ b/tests/all/wast.rs
@@ -32,7 +32,7 @@ fn run_wast(wast: &str, strategy: Strategy) -> anyhow::Result<()> {
     // OOM killer.
     //
     // Locally testing this out this drops QEMU's memory usage running this
-    // tests suite from 10GiB to 600MiB. Prevoiusly we saw that crossing the
+    // tests suite from 10GiB to 600MiB. Previously we saw that crossing the
     // 10GiB threshold caused our processes to get OOM killed on CI.
     if std::env::var("WASMTIME_TEST_NO_HOG_MEMORY").is_ok() {
         cfg.static_memory_maximum_size(0);

